### PR TITLE
Refactoring Xeno Ability Buttons

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
Refactors xeno ability buttons to be more modular and consistent

## Why It's Good For The Game
makes behaviour more consistent and makes the code more maintainable

## Changelog
:cl:
tweak: Regurgitate will disabled if you have nothing to regurgitate
tweak: Many xeno ability buttons will now red/grey out when you can't use them.
tweak: Abilities on cooldown or that you lack plasma for are now selectable to allow quick switching
tweak: Abilities on cooldown now show a moving clock on top of the icon
tweak: Abilities on cooldown or you lack plasma for now show the green outline showing they're selected
/:cl:

- [x] Boiler bombard tested
- [x] Warrior abilities tested
- [x] Defender abilities tested
- [x] Drone abilities tested
- [x] Spitter/sent abilities tested

https://gyazo.com/95313c1c115a4b95031e83798e492377

### CODE REVIEW

File: code/datums/action.dm [hunk 1]
```diff
@@ -106,28 +106,68 @@
...
 /datum/action/xeno_action/New(Target)
 	..()
 	button.overlays += image('icons/mob/actions.dmi', button, action_icon_state)
 
 /datum/action/xeno_action/can_use_action()
 	var/mob/living/carbon/Xenomorph/X = owner
-	if(X && !X.incapacitated() && !X.lying && !X.buckled && X.plasma_stored >= plasma_cost && !X.stagger)
-		return TRUE
+	if(!X)
+		return FALSE
```
Comment (Rohesie) [inline code/datums/action.dm]: Add a `stack_trace("[src] action attempted to check can_use_action() without an owner")`

**Comment by spookydonut** [comment]
This is close to completion, reviews can begin

File: code/__DEFINES/mobs.dm [hunk 2]
```diff
@@ -564,7 +564,11 @@ GLOBAL_LIST_INIT(xenoupgradetiers, list(XENO_UPGRADE_BASETYPE, XENO_UPGRADE_INVA
 #define CRUSHER_CHARGE_RAZORWIRE_MULTI	100
 #define CRUSHER_CHARGE_TANK_MULTI		100
 
-#define CRUSHER_STOMP_UPGRADE_BONUS ( 1 + upgrade_as_number() * 0.05 )
+#define CHARGE_TURFS_TO_CHARGE			5		//Amount of turfs to build up before a charge begins
+#define CHARGE_SPEED_BUILDUP			0.15 	//POSITIVE amount of speed built up during a charge each step
```
Comment (DominikPanic) [inline code/__DEFINES/mobs.dm]: This is duplicated in #1233

File: code/game/turfs/turf.dm [hunk 3]
```diff
@@ -359,23 +362,50 @@
...
-	else
-		return TRUE
 
 
 /turf/closed/wall/is_weedable()
-	return TRUE //so we can spawn weeds on the walls
-
+	return !is_type_in_typecache(get_area(src), GLOB.unweedable_areas) //so we can spawn weeds on the walls
+
+/turf/proc/check_alien_construction(var/mob/living/L)
```
Comment (DominikPanic) [inline code/game/turfs/turf.dm]:
```suggestion
/turf/proc/check_alien_construction(mob/living/L)
```

File: code/modules/mob/living/carbon/xenomorph/castes/hunter/abilities_hunter.dm [hunk 4]
```diff
@@ -72,54 +72,24 @@
...
-
-
-	return TRUE
+/datum/action/xeno_action/activable/pounce/hunter
+	plasma_cost = 20
+	range = 7
+
+/datum/action/xeno_action/activable/pounce/hunter/prepare_to_pounce()
+	. = ..()
+	if(owner.m_intent == "walk") //Hunter that is currently using its stealth ability, need to unstealth him
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/castes/hunter/abilities_hunter.dm]:
```/////////////////MOVE DEFINES//////////////////////
#define MOVE_INTENT_WALK        1
#define MOVE_INTENT_RUN         2

File: code/modules/mob/living/carbon/xenomorph/castes/praetorian/abilities_praetorian.dm [hunk 5]
```diff
@@ -1,59 +1,151 @@
...
 /mob/living/carbon/Xenomorph/proc/speed_increase(var/amount)
 	speed -= amount
 
-/mob/living/carbon/Xenomorph/proc/acid_spray_cooldown_finished()
-	used_acid_spray = FALSE
-	to_chat(src, "<span class='notice'>You have produced enough acid to spray again.</span>")
+GLOBAL_LIST_INIT(acid_spray_hit, typecacheof(list(/obj/structure/barricade, /obj/vehicle/multitile/root/cm_armored, /obj/structure/razorwire)))
+
+/datum/action/xeno_action/activable/spray_acid/cone/proc/do_acid_spray_cone(turf/T, range)
+	set waitfor = 0
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/castes/praetorian/abilities_praetorian.dm]:
```suggestion
	set waitfor = FALSE
```

File: code/__DEFINES/mobs.dm (hunk 2 above)
Comment (spookydonut) [reply code/__DEFINES/mobs.dm]: Yeah will resolve conflicts with it after that's merged.

**Comment by Rohesie** [comment]
I can see what you said on the checks now. Much, much cleaner now.

File: code/datums/action.dm [hunk 6]
```diff
@@ -160,8 +272,9 @@
 		button.icon_state = "template_on"
 		X.selected_ability = src
 		X.selected_ability.on_activation()
-	if(plasma_cost)
+	if(use_plasma_on_activate && plasma_cost)
```
Comment (Rohesie) [inline code/datums/action.dm]: Nothing is modifying `use_plasma_on_activate ` so it will always be null.

File: code/datums/action.dm [hunk 7]
```diff
@@ -139,13 +251,13 @@
 
 
 /datum/action/xeno_action/activable
-	var/ability_name
+	var/use_plasma_on_activate
```
Comment (Rohesie) [inline code/datums/action.dm]: This doesn't seem to be modified anywhere, and it's a flag now.

File: code/datums/action.dm [hunk 8]
```diff
@@ -173,6 +286,21 @@
 /datum/action/xeno_action/activable/proc/use_ability(atom/A)
 	return
 
+//override this 
+/datum/action/xeno_action/activable/proc/can_use_ability(atom/A, silent = FALSE, override_flags)
+	var/mob/living/carbon/Xenomorph/X = owner
```
Comment (Rohesie) [inline code/datums/action.dm]: Type cast `owner`? Or will anything but a xeno be able to use `xeno_action`?

File: code/game/turfs/turf.dm [hunk 9]
```diff
@@ -359,23 +362,50 @@
...
 
 /turf/closed/wall/is_weedable()
-	return TRUE //so we can spawn weeds on the walls
-
+	return !is_type_in_typecache(get_area(src), GLOB.unweedable_areas) //so we can spawn weeds on the walls
+
+/turf/proc/check_alien_construction(mob/living/L)
+	var/has_obstacle
+	for(var/obj/O in contents)
+		if(istype(O, /obj/item/clothing/mask/facehugger))
```
Comment (Rohesie) [inline code/game/turfs/turf.dm]: I wonder here what's more efficient, if redefined procs or these multiple checks.
@MrStonedOne seemed to suggest proc overhead was worse, while @ninjanomnom suggested multiple checks to be less desirable.

File: code/datums/action.dm (hunk 7 above)
Comment (spookydonut) [reply code/datums/action.dm]: yeah

Comment (spookydonut) [reply code/datums/action.dm]: nothing seems to rely on the proc using plasma so i'll just remove that part.

File: code/modules/mob/living/carbon/xenomorph/castes/spitter/abilities_spitter.dm [hunk 10]
```diff
@@ -1 +1,90 @@
...
+	playsound(X.loc, 'sound/effects/refill.ogg', 50, 1)
+	X.visible_message("<span class='xenowarning'>\The [X] spews forth a virulent spray of acid!</span>", \
+	"<span class='xenowarning'>You spew forth a spray of acid!</span>", null, 5)
+	var/turflist = getline(X, target)
+	spray_turfs(turflist)
+	add_cooldown()
+
+
+/datum/action/xeno_action/activable/spray_acid/line/proc/spray_turfs(list/turflist)
+	set waitfor = 0
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/castes/spitter/abilities_spitter.dm]:
```suggestion
	set waitfor = FALSE
```

**Comment by Rohesie** [comment]
Set the testing label once conflict is resolved. Maybe comment here so we get the update.

**Comment by Rohesie** [comment]
Moving screech from mob proc to datum proc broke its logic. It no longer works.

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 11]
```diff
@@ -110,140 +110,114 @@
...
+	add_cooldown()
+
+	playsound(X.loc, 'sound/voice/alien_queen_screech.ogg', 75, 0)
+	X.visible_message("<span class='xenohighdanger'>\The [X] emits an ear-splitting guttural roar!</span>")
 	round_statistics.queen_screech++
-	create_shriekwave() //Adds the visual effect. Wom wom wom
+	X.create_shriekwave() //Adds the visual effect. Wom wom wom
 	//stop_momentum(charge_dir) //Screech kills a charge
 
 	for(var/mob/living/L in range(world.view))
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm]:
```suggestion
	for(var/mob/living/L in X.range(world.view))
```

Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm]: range is a global proc not an atom proc.

**Pull request merged**